### PR TITLE
add registry and macros

### PR DESCRIPTION
**Pull request by BusyJay**

### CODE REVIEW

File: tests/tests.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        fail_point!("sleep");
+    };
+    let timer = Instant::now();
+    f();
+    assert!(timer.elapsed() < Duration::from_millis(1000));
+
+    let timer = Instant::now();
+    fail::cfg("tests::sleep", "sleep(1000)").unwrap();
+    f();
+    assert!(timer.elapsed() > Duration::from_millis(1000));
```
Comment (siddontang) [inline tests/tests.rs]: if our test run quickly, maybe we should use `>=` to check here.

File: src/lib.rs [hunk 2]
```diff
@@ -256,6 +257,98 @@ impl FailPoint {
...
+        p.set_actions(vec![]);
+    }
+}
+
+fn set(
+    registry: &mut HashMap<String, Arc<FailPoint>>,
+    name: String,
+    orders: &str,
+) -> Result<(), String> {
+    let actions = try!(orders.split("->").map(Action::from_str).collect());
```
Comment (siddontang) [inline src/lib.rs]: should add a comment for `->`

File: src/lib.rs [hunk 3]
```diff
@@ -256,6 +257,98 @@ impl FailPoint {
...
+struct FailPointRegistry {
+    // TODO: remove rwlock or store *mut FailPoint
+    registry: RwLock<HashMap<String, Arc<FailPoint>>>,
+}
+
+lazy_static! {
+    static ref REGISTRY: FailPointRegistry = FailPointRegistry::default();
+}
+
+pub fn setup() {
```
Comment (siddontang) [inline src/lib.rs]: should add comments for all public APIs.

**Comment by siddontang** [comment]
PTAL @overvenus

File: tests/tests.rs (hunk 1 above)
Comment (BusyJay) [reply tests/tests.rs]: Actually it's impossible. There are many instructions before sleeping which may consumes 100ns+ time.

File: src/lib.rs (hunk 3 above)
Comment (BusyJay) [reply src/lib.rs]: I will fix all docs in another pr when all things are settled down.

**Comment by siddontang** [comment]
PTAL @overvenus

File: src/lib.rs [hunk 4]
```diff
@@ -256,6 +257,99 @@ impl FailPoint {
...
+#[macro_export]
+macro_rules! fail_point {
+    ($name:expr, $e:expr) => {{
+        let name = concat!(module_path!(), "::", $name);
+        if let Some(res) = $crate::eval(name, $e) {
+            return res;
+        }
+    }};
+    ($name:expr) => {{
+        fail_point!($name, |_| {});
```
Comment (overvenus) [inline src/lib.rs]: Does it only fit in functions that return nothing? `Fn<Args>`

File: src/lib.rs [hunk 5]
```diff
@@ -256,6 +257,99 @@ impl FailPoint {
...
+pub fn teardown() {
+    let mut registry = REGISTRY.registry.write().unwrap();
+    for (_, p) in &*registry {
+        // wake up all pause failpoint.
+        p.set_actions(vec![]);
+    }
+    registry.clear();
+}
+
+pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
```
Comment (overvenus) [inline src/lib.rs]: Please add `#[doc(hidden)] ` if it should not be invoked by users.

File: src/lib.rs (hunk 4 above)
Comment (BusyJay) [reply src/lib.rs]: Yes.

File: src/lib.rs [hunk 6]
```diff
@@ -256,6 +257,99 @@ impl FailPoint {
...
+    let p = registry
+        .entry(name)
+        .or_insert_with(|| Arc::new(FailPoint::new()));
+    p.set_actions(actions);
+    Ok(())
+}
+
+#[macro_export]
+macro_rules! fail_point {
+    ($name:expr, $e:expr) => {{
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/lib.rs]: It will be done once features is introduced.

**Comment by overvenus** [comment]
LGTM

**Pull request merged**